### PR TITLE
Feat: add if in workflow

### LINKED ISSUE #3825: [Feature] workflow support  get application status for notification after `apply-application`
**Is your feature request related to a problem? Please describe.**

no

In my CD workflow:
- use  `apply-application`, I want to know application status for notification
  - if application health, send notification success message
  - if some error or application unhealthy within a certain time , send notification  failed message. 

**Describe the solution you'd like**

no

**Describe alternatives you've considered**

no

**Additional context**

no

### PR DESCRIPTION
**Pull request by FogDong**
### Description of your changes

This PR adds `if` in both workflow steps and substeps.

In the KubeVela v1.3, if the step failed in workflow, after ten times retries, the workflow will suspend automatically.
After this PR is merged, if one step failed after retries, its reason is `FailedAfterRetries`, and the workflow will try to execute the following steps. For the steps that can't satisfy the execution conditions, their phase will be `skipped`. After all steps have been executed, the workflow's phase is `terminated`.

> Note that you can use --enable-suspend-failed-workflow flag to open the feature of suspend the failed workflow automatically, but with the feature, you can not use `if`.

Here's an example.

The workflow step spec is like this:
```yaml
  workflow:
    steps:
    - name: webhook-err
      type: webhook-err
    - name: notification
      type: notification
      if: always
    - name: notification2
      type: notification
```

And the status is like:
```yaml
    status: workflowTerminated
    workflow:
      finished: true
      message: The workflow terminates automatically because the failed times of steps
        have reached the limit
      mode: StepByStep
      steps:
      - firstExecuteTime: "2022-05-27T02:57:14Z"
        id: adkxovfvvi
        lastExecuteTime: "2022-05-27T02:58:06Z"
        message: 'step webhook: step http: run step(provider=http,do=do): invalid
          string argument, webhook.http.url: field "str" is optional'
        name: webhook-err
        phase: failed
        reason: FailedAfterRetries
        type: webhook-err
      - firstExecuteTime: "2022-05-27T02:58:06Z"
        id: nwwmr9ysa6
        lastExecuteTime: "2022-05-27T02:58:06Z"
        name: notification
        phase: succeeded
        type: notification
      - firstExecuteTime: "2022-05-27T02:58:06Z"
        id: pohtl28pwz
        lastExecuteTime: "2022-05-27T02:58:06Z"
        name: notification2
        phase: skipped
        reason: Skip
        type: notification
      suspend: false
      terminated: true
```

If you try to terminate a running workflow manually, the running step's phase will be `failed`, and its reason is `Terminated`. e.g.

```yaml
    status: workflowTerminated
    workflow:
      finished: true
      message: terminated
      mode: StepByStep
      steps:
      - firstExecuteTime: "2022-05-27T03:01:51Z"
        id: ijv8go9bsa
        lastExecuteTime: "2022-05-27T03:01:55Z"
        message: wait healthy
        name: apply
        phase: failed
        reason: Terminate
        type: apply-component
      - firstExecuteTime: "2022-05-27T03:01:56Z"
        id: 8jx3ke4kix
        lastExecuteTime: "2022-05-27T03:01:56Z"
        name: notification
       // this step is executed successfully because it's `if` is `always`.
        phase: succeeded
        type: notification
      - firstExecuteTime: "2022-05-27T03:01:56Z"
        id: 8gpmipwkz3
        lastExecuteTime: "2022-05-27T03:01:56Z"
        name: notification2
        phase: skipped
        reason: Skip
        type: notification
      suspend: false
      terminated: true
```

Fixes #3825

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

**Comment by FogDong** [comment]
Hold this PR until #3772 is merged since there'll be many conflicts that need to be solved.

File: pkg/workflow/tasks/custom/task.go [hunk 1]
```diff
@@ -441,3 +465,61 @@ func NewTaskLoader(lt LoadTaskTemplate, pkgDiscover *packages.PackageDiscover, h
...
+		for _, depend := range options.DependsOn {
+			if status, ok := options.StepStatus[depend]; ok {
+				if status.Phase == common.WorkflowStepPhaseSkipped || status.Phase == common.WorkflowStepPhaseFailedAfterRetries {
+					return true
+				}
+			}
+		}
+		return false
+	default:
+		// TODO:(fog)
```
Comment (wonderflow) [inline pkg/workflow/tasks/custom/task.go:495]: what should we todo here?

Comment (wonderflow) [reply pkg/workflow/tasks/custom/task.go:495]: you should add this into the code.

File: pkg/workflow/tasks/custom/task.go [hunk 2]
```diff
@@ -441,3 +463,49 @@ func NewTaskLoader(lt LoadTaskTemplate, pkgDiscover *packages.PackageDiscover, h
 		logLevel: logLevel,
 	}
 }
+
+// SkipOptions is the options of skip task runner
+type SkipOptions struct {
+	If             string
+	DependsOnPhase common.WorkflowStepPhase
+}
```
Comment (leejanee) [inline pkg/workflow/tasks/custom/task.go:471]: 这个定义是否意味着跳过的判断依据 要么是always要么是某个指定步骤的执行结果？

**Review by leejanee — COMMENTED** [review]
Does substep also support `if`?

File: pkg/workflow/tasks/custom/task.go (hunk 2 above)
Comment (FogDong) [reply pkg/workflow/tasks/custom/task.go:471]: 如果 if 为空，会根据这个步骤之前的执行结果来判断是否执行（失败则跳过，成功则执行）

Comment (barnettZQG) [reply pkg/workflow/tasks/custom/task.go:473]: 是之前的所有步骤吗？

Comment (FogDong) [reply pkg/workflow/tasks/custom/task.go:473]: 是的，如果之前的步骤有一个 failed 了，且当前步骤没有 if-always，那么它也会被跳过

File: pkg/apiserver/domain/service/workflow.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ import (
 	"github.com/oam-dev/kubevela/pkg/oam/util"
 	utils2 "github.com/oam-dev/kubevela/pkg/utils"
 	"github.com/oam-dev/kubevela/pkg/utils/apply"
+	"github.com/oam-dev/kubevela/references/cli"
 )
```
Comment (Somefive) [inline pkg/apiserver/domain/service/workflow.go:47]: Strongly discourage importing reference/ package in pkg/ package. Related functions should be moved in pkg/.

Comment (wonderflow) [reply pkg/apiserver/domain/service/workflow.go:47]: exactly yes!

Comment (wonderflow) [reply pkg/apiserver/domain/service/workflow.go:47]: maybe we will migrate the cli into a separate project some day

File: charts/vela-minimal/values.yaml [hunk 4]
```diff
@@ -38,10 +38,12 @@ dependCheckWait: 30s
 
 ## @section KubeVela workflow parameters
 
+## @param workflow.enableSuspendFailedWorkflow Enable suspend failed workflow
 ## @param workflow.backoff.maxTime.waitState The max backoff time of workflow in a wait condition
 ## @param workflow.backoff.maxTime.failedState The max backoff time of workflow in a failed condition
 ## @param workflow.step.errorRetryTimes The max retry times of a failed workflow step
 workflow:
+  enableSuspendFailedWorkflow: false
```
Comment (wonderflow) [inline charts/vela-minimal/values.yaml:46]:
```suggestion
  enableSuspendOnFailure: false
```

File: pkg/apiserver/domain/service/workflow.go [hunk 5]
```diff
@@ -588,6 +588,42 @@ func (w *workflowServiceImpl) TerminateRecord(ctx context.Context, appModel *mod
...
+				}
+			case common.WorkflowStepPhaseRunning:
+				steps[i].SubStepsStatus[j].Phase = common.WorkflowStepPhaseFailed
+				steps[i].SubStepsStatus[j].Reason = custom.StatusReasonTerminate
+			default:
+			}
+		}
+	}
+
+	if err := kubecli.Status().Patch(context.TODO(), app, client.Merge); err != nil {
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/workflow.go:621]: let's pass the context in

File: pkg/features/controller_features.go [hunk 6]
```diff
@@ -33,6 +33,8 @@ const (
 	DeprecatedObjectLabelSelector featuregate.Feature = "DeprecatedObjectLabelSelector"
 	// LegacyResourceTrackerGC enable the gc of legacy resource tracker in managed clusters
 	LegacyResourceTrackerGC featuregate.Feature = "LegacyResourceTrackerGC"
+	// EnableSuspendFailedWorkflow enable suspend failed workflow
+	EnableSuspendFailedWorkflow featuregate.Feature = "EnableSuspendFailedWorkflow"
```
Comment (wonderflow) [inline pkg/features/controller_features.go:37]:
```suggestion
	EnableSuspendFailedWorkflow featuregate.Feature = "EnableSuspendOnFailure"
```

File: pkg/workflow/hooks/data_passing.go [hunk 7]
```diff
@@ -66,13 +56,6 @@ func Output(ctx wfContext.Context, taskValue *value.Value, step v1beta1.Workflow
 			if err := json.Unmarshal(js, &o); err != nil {
 				return err
 			}
-			ready, err := value.NewValue(`true`, nil, "")
-			if err != nil {
-				return err
-			}
-			if err := ctx.SetVar(ready, ReadyComponent, o.Name); err != nil {
-				return err
-			}
```
Comment (wonderflow) [inline pkg/workflow/hooks/data_passing.go:75]: why don't we need this here?

Comment (FogDong) [reply pkg/workflow/hooks/data_passing.go:75]: I pass all the step status as a map to the `Pending` so we can get the status from the map directly, there's no need that we store another copy of ready step status in ctx.

File: pkg/workflow/workflow.go [hunk 8]
```diff
@@ -190,23 +155,110 @@ func (w *workflow) ExecuteSteps(ctx monitorContext.Context, appRev *oamcore.Appl
 	}
 
 	e.checkWorkflowStatusMessage(wfStatus)
+	fmt.Println(99999, e.status.Message)
```
Comment (wonderflow) [inline pkg/workflow/workflow.go:158]: debug info

File: pkg/workflow/workflow.go [hunk 9]
```diff
@@ -190,23 +155,110 @@ func (w *workflow) ExecuteSteps(ctx monitorContext.Context, appRev *oamcore.Appl
...
+	if allTasksSucceeded {
 		wfStatus.Message = string(common.WorkflowStateSucceeded)
 		return common.WorkflowStateSucceeded, nil
 	}
 	wfStatus.Message = string(common.WorkflowStateExecuting)
 	return common.WorkflowStateExecuting, nil
 }
 
+func (w *workflow) restartWorkflow(ctx monitorContext.Context, revAndSpecHash string) (common.WorkflowState, error) {
+	ctx.Info("Restart Workflow")
```
Comment (wonderflow) [inline pkg/workflow/workflow.go:185]: debug info?

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 10]
```diff
@@ -239,6 +241,9 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		if err := r.doWorkflowFinish(app, wf); err != nil {
 			return r.endWithNegativeCondition(ctx, app, condition.ErrorCondition(common.WorkflowCondition.String(), errors.WithMessage(err, "DoWorkflowFinish")), common.ApplicationRunningWorkflow)
 		}
+		if !workflow.IsFailedAfterRetry(app) {
+			r.stateKeep(logCtx, handler, app)
```
Comment (FogDong) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:245]: delete state keep here.

**Review by wonderflow — APPROVED** [review]
LGTM after nits

**Review by Somefive — APPROVED** [review]

File: pkg/workflow/tasks/custom/task.go [hunk 11]
```diff
@@ -441,3 +465,52 @@ func NewTaskLoader(lt LoadTaskTemplate, pkgDiscover *packages.PackageDiscover, h
...
+// IsStepFinish will decide whether step is finish.
+func IsStepFinish(phase common.WorkflowStepPhase, reason string) bool {
+	if feature.DefaultMutableFeatureGate.Enabled(features.EnableSuspendOnFailure) {
+		return phase == common.WorkflowStepPhaseSucceeded
+	}
+	if phase == common.WorkflowStepPhaseFailed {
+		return reason == StatusReasonTerminate || reason == StatusReasonFailedAfterRetries
+	}
+	return phase == common.WorkflowStepPhaseSucceeded || phase == common.WorkflowStepPhaseSkipped
+}
```
Comment (leejanee) [inline pkg/workflow/tasks/custom/task.go:516]: Doesn't it mean that step  is finished when it‘s phase is not `running`?

**Review by leejanee — APPROVED** [review]
LGTM

**Pull request merged**